dev/PushNotifications/PushNotificationsLongRunningTask/platform.cpp

```diff
@@ -0,0 +1,185 @@
+﻿#include "pch.h"
+
+#include "platform.h"
+#include "platformfactory.h"
+#include <FrameworkUdk/PushNotifications.h>
+
+using namespace winrt::Windows;
+
+void NotificationsLongRunningPlatformImpl::Initialize()
+{
+    auto lock = m_lock.lock_exclusive();
+    THROW_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    if (m_initialized)
+    {
+        return;
+    }
+
+    // Schedule event signaling to happen in 5 seconds.
+    // This is in case we later realize there are no apps to be tracked in the LRP.
+    m_lifetimeManager.Setup();
+
+    m_storage = Storage::ApplicationData::Current().LocalSettings().CreateContainer(
+        L"LRP", Storage::ApplicationDataCreateDisposition::Always);
+
+    m_foregroundSinkManager = std::make_shared<ForegroundSinkManager>();
+
+    auto fullTrustApps = GetFullTrustApps();
+
+    if (!fullTrustApps.empty())
+    {
+        // We have at least one app that could receive notifications.
+        // Cancel the timer to persist the LRP.
+        m_lifetimeManager.Cancel();
+        m_notificationListenerManager.Initialize(m_foregroundSinkManager, fullTrustApps);
+    }
+
+    m_initialized = true;
+}
+
+void NotificationsLongRunningPlatformImpl::Shutdown() noexcept
+{
+    auto lock = m_lock.lock_exclusive();
+    if (m_shutdown)
+    {
+        return;
+    }
+
+    m_shutdown = true;
+}
+
+void NotificationsLongRunningPlatformImpl::WaitForLifetimeEvent()
+{
+    m_lifetimeManager.Wait();
+}
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterFullTrustApplication(
+    _In_ PCWSTR processName, GUID remoteId, _Out_ PWSTR* appId) noexcept try
+{
+    auto lock = m_lock.lock_exclusive();
+    THROW_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    std::wstring appIdentifier = GetAppIdentifier(processName);
+    THROW_IF_FAILED(PushNotifications_RegisterFullTrustApplication(appIdentifier.c_str(), remoteId));
+
+    wil::unique_cotaskmem_string outAppId = wil::make_unique_string<wil::unique_cotaskmem_string>(appIdentifier.c_str());
+    *appId = outAppId.release();
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::RegisterActivator(_In_ PCWSTR processName) noexcept try
+{
+    auto lock = m_lock.lock_shared();
+    THROW_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    std::wstring appId = GetAppIdentifier(processName);
+    m_notificationListenerManager.AddListener(appId, processName);
+
+    m_lifetimeManager.Cancel();
+
+    return S_OK;
+}
+CATCH_RETURN()
+
+STDMETHODIMP_(HRESULT __stdcall) NotificationsLongRunningPlatformImpl::UnregisterActivator(_In_ PCWSTR processName) noexcept try
+{
+    auto lock = m_lock.lock_shared();
+    THROW_HR_IF(WPN_E_PLATFORM_UNAVAILABLE, m_shutdown);
+
+    std::wstring appId = GetAppIdentifier(processName);
+    m_notificationListenerManager.RemoveListener(appId);
+    m_foregroundSinkManager->Remove(appId);
```

Comment:
This is a bug. It would throw if there are no foreground sinks. The fix would be to check if a fg sink exists for that appId before removing it here.

Reply:
map::erase doesn't throw. It would just return 0 but we aren't doing anything with the return type.